compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2429,6 +2429,80 @@ TR::Register *OMR::X86::TreeEvaluator::andORStringEvaluator(TR::Node *node, TR::
    return resultReg;
    }
 
+TR::Register *OMR::X86::TreeEvaluator::OverflowCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   /*
+    *overflowCHK
+    *   =>child1(Operation)
+    *   =>child2(Operand1)
+    *   =>child3(Operand2)
+    */
+   bool firstChildEvaluatedAlready = node->getFirstChild()->getRegister()? true : false;
+   cg->evaluate(node->getFirstChild());
+   TR::Block *overflowCatchBlock = NULL;
+   TR::Instruction *prevInst = cg->comp()->getAppendInstruction();
+   // we need to do the operation again for the following cases:
+   // 1. the operation was somehow optimized to other forms, like LEA for example
+   // 2. the child was evaluated already under a different treetop
+   if (!(prevInst->getOpCode()).modifiesOverflowFlag() || firstChildEvaluatedAlready ) 
+      {
+      TR::Register *reg1; 
+      TR::Register *reg2 = cg->evaluate(node->getThirdChild());
+      TR_X86OpCodes op;
+      bool nodeIs64Bit; 
+      switch (node->getOverflowCHKInfo())
+         {
+         case TR::ladd:
+            nodeIs64Bit= true;
+            TR_ASSERT(TR::Compiler->target.is64Bit(), "overflowCHK for long add is only supported on x86 64 bits platform!!");
+            op = ADD8RegReg;
+            break;
+         case TR::iadd:
+            nodeIs64Bit = false;
+            op = ADD4RegReg;
+            break;
+         case TR::sadd:
+            nodeIs64Bit = false;
+            op = ADD2RegReg;
+            break;
+         case TR::badd:
+            nodeIs64Bit = false;
+            op = ADD1RegReg;
+            break;
+         default:
+            TR_ASSERT(0 , "unsupported OverflowCHK opcode on node %p\n", node);
+         }
+      reg1 = intOrLongClobberEvaluate(node->getSecondChild(), nodeIs64Bit, cg);
+      generateRegRegInstruction(op, node, reg1, reg2, cg);
+      cg->stopUsingRegister(reg1);
+      }
+
+   TR::list<TR::CFGEdge*> excepSucc =cg->getCurrentEvaluationTreeTop()->getEnclosingBlock()->getExceptionSuccessors();
+   for (auto e = excepSucc.begin(); e != excepSucc.end(); ++e)
+      {
+      TR::Block *dest = toBlock((*e)->getTo());
+      if (dest->getCatchBlockExtension()->_catchType == TR::Block::CanCatchOverflowCheck)
+            overflowCatchBlock = dest;
+      }
+   TR_ASSERT(overflowCatchBlock != NULL, "OverflowChk node %p doesn't have overflow catch block\n", node);
```

Comment:
I still think this needs to be moved up in the method to ensure we are not part way through evaluation when we decide there is a problem.